engine/runtime/src/main/java/org/enso/interpreter/runtime/data/ArraySlice.java

```diff
@@ -0,0 +1,113 @@
+package org.enso.interpreter.runtime.data;
+
+import com.oracle.truffle.api.CompilerDirectives;
+import com.oracle.truffle.api.dsl.Cached;
+import com.oracle.truffle.api.interop.*;
+import com.oracle.truffle.api.library.CachedLibrary;
+import com.oracle.truffle.api.library.ExportLibrary;
+import com.oracle.truffle.api.library.ExportMessage;
+import org.enso.interpreter.node.expression.builtin.interop.syntax.HostValueToEnsoNode;
+
+@ExportLibrary(InteropLibrary.class)
+public class ArraySlice implements TruffleObject {
+  private final Object storage;
+  private final long start;
+  private final long end;
+
+  public ArraySlice(Object storage, long start, long end) {
```

Comment:
Who's supposed to call this constructor? Just `Vector.slice` from the same package? Then make it package private. Possibly also make the whole class package private.